File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/service/event/NotificationEventListener.java
package com.github.arhor.simple.expense.tracker.service.event;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;

import java.io.IOException;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.github.arhor.simple.expense.tracker.service.NotificationService;

@Slf4j
@Service
@RequiredArgsConstructor(onConstructor_ = @Autowired)
public class NotificationEventListener {

    private final NotificationService service;

    @Async
    @Transactional
    @EventListener(NotificationEvent.class)
    public void handleNotificationEvent(final NotificationEvent event) throws IOException {
        val userId = event.userId();
        val senderId = event.senderId();
        val notification = event.notification();

        log.debug("Handled notification event: {}, for the user: {}, from user: {}", notification, userId, senderId);

        service.sendNotification(senderId, userId, notification);
    }
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/service/event/NotificationEventListener.kt
package com.github.arhor.simple.expense.tracker.service.event

import com.github.arhor.simple.expense.tracker.service.NotificationService
import org.slf4j.LoggerFactory
import org.springframework.context.event.EventListener
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service

@Service
class NotificationEventListener(private val service: NotificationService) {

    @Async
    @EventListener
    fun handleNotificationEvent(event: NotificationEvent) {
        val (sourceUserId, targetUserId, notification) = event

        log.debug(
            "Handled notification event: from user: {}, to user: {}",
            sourceUserId,
            targetUserId,
        )
        service.sendNotification(sourceUserId, targetUserId, notification)
    }

    companion object {
        private val log = LoggerFactory.getLogger(NotificationEventListener::class.java)
    }
}
